pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,364 @@
+	sfs.wg.Wait()
+
+	sfs.isShutdown.Store(true)
+	if sfs.shutdownCallback != nil {
+		sfs.shutdownCallback()
+	}
+	sfs.log.Info("state fetcher shutdown")
+}
+
+func (sfs *Service) run() {
+	var (
+		syncHeight   uint32
+		expectedRoot util.Uint256
+	)
+
+	defer func() {
+		sfs.isActive.Store(false)
+		sfs.isShutdown.Store(true)
+		sfs.basic.CtxCancel()
+		sfs.wg.Done()
+		if sfs.shutdownCallback != nil {
+			sfs.shutdownCallback()
+		}
+		sfs.log.Info("state fetcher stopped")
+	}()
+	reader, err := sfs.getStateObject()
+	if err != nil {
+		sfs.log.Error("failed to get state object", zap.Error(err))
+		return
+	}
+	defer func() {
+		if err = reader.Close(); err != nil {
+			sfs.log.Warn("failed to close reader", zap.Error(err))
+		}
+	}()
+
+	var (
+		lastKey = sfs.chain.GetLastStoredKey()
+		kvChan  = make(chan kvPair, 1)
+		batch   = make([]statesync.KeyValue, 0, sfs.cfg.KeyValueBatchSize)
+	)
+	go func() {
+		defer close(kvChan)
+		br := gio.NewBinReaderFromIO(reader)
+
+		version := br.ReadB()
+		if version != 0 || br.Err != nil {
+			kvChan <- kvPair{err: fmt.Errorf("invalid state object version: %d, error: %w", version, br.Err)}
+			return
+		}
+		_ = br.ReadU32LE() // Skip network magic check.
+		syncHeight = br.ReadU32LE()
+		br.ReadBytes(expectedRoot[:])
+		if br.Err != nil {
+			kvChan <- kvPair{err: fmt.Errorf("failed to read state root: %w", br.Err)}
+			return
+		}
+		sfs.log.Info("state object",
+			zap.String("root", expectedRoot.StringLE()),
+			zap.Uint32("height", syncHeight))
+
+		for {
+			select {
+			case <-sfs.basic.Ctx.Done():
+				kvChan <- kvPair{err: sfs.basic.Ctx.Err()}
+				return
+			default:
+				key := br.ReadVarBytes()
+				if br.Err != nil {
+					kvChan <- kvPair{err: br.Err}
+					return
+				}
+				value := br.ReadVarBytes()
+				if br.Err != nil {
+					kvChan <- kvPair{err: br.Err}
+					return
+				}
+
+				if len(lastKey) > 0 && bytes.Compare(key, lastKey) <= 0 {
+					sfs.log.Debug("skipping key", zap.String("key", hex.EncodeToString(key)))
+					continue
+				}
+				kvChan <- kvPair{statesync.KeyValue{Key: key, Value: value}, nil}
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-sfs.basic.Ctx.Done():
+			sfs.log.Debug("shutdown requested")
+			if len(batch) > 0 {
+				if err = sfs.chain.AddContractStorageData(batch, syncHeight, expectedRoot); err != nil {
+					sfs.log.Error("failed to add storage pair in shutdown", zap.Error(err))
+					return
+				}
+			}
+			return
+		case pair, ok := <-kvChan:
+			if !ok || pair.err != nil {
+				if pair.err != nil && !errors.Is(pair.err, io.EOF) {
+					sfs.log.Error("failed to read state stream", zap.Error(pair.err))
+					return
+				}
+				if len(batch) > 0 {
+					if err = sfs.chain.AddContractStorageData(batch, syncHeight, expectedRoot); err != nil {
+						sfs.log.Error("failed to add storage pair in final batch", zap.Error(err))
+						return
+					}
+				}
+				sfs.log.Info("state fetch completed", zap.Uint32("height", syncHeight))
+				return
```

Comment:
Remove this log. Once you integrate the `exiter` function, it will notify the user about service exit.